language/move-vm/types/src/values/values_impl.rs

```diff
@@ -607,46 +568,24 @@ impl Value {
  *
  **************************************************************************************/
 
-impl ContainerRef {
-    fn read_ref(self) -> PartialVMResult<Value> {
-        Ok(Value(ValueImpl::Container(self.container().copy_value()?)))
-    }
-}
-
-impl IndexedRef {
-    fn read_ref(self) -> PartialVMResult<Value> {
-        use Container::*;
-
-        let res = match &*self.container_ref.container() {
-            Locals(r) | Vec(r) | Struct(r) => r.borrow()[self.idx].copy_value()?,
-            VecU8(r) => ValueImpl::U8(r.borrow()[self.idx]),
-            VecU64(r) => ValueImpl::U64(r.borrow()[self.idx]),
-            VecU128(r) => ValueImpl::U128(r.borrow()[self.idx]),
-            VecBool(r) => ValueImpl::Bool(r.borrow()[self.idx]),
-            VecAddress(r) => ValueImpl::Address(r.borrow()[self.idx]),
-        };
-
-        Ok(Value(res))
-    }
-}
-
 impl ReferenceImpl {
-    fn read_ref(self) -> PartialVMResult<Value> {
-        match self {
-            Self::ContainerRef(r) => r.read_ref(),
-            Self::IndexedRef(r) => r.read_ref(),
-        }
-    }
-}
-
-impl StructRef {
-    pub fn read_ref(self) -> PartialVMResult<Value> {
-        self.0.read_ref()
+    fn read_ref(self) -> Value {
+        let v = unsafe {
+            match self.checked_ref() {
+                CheckedRef::U8(u) => ValueImpl::U8(*u),
+                CheckedRef::U64(u) => ValueImpl::U64(*u),
+                CheckedRef::U128(u) => ValueImpl::U128(*u),
+                CheckedRef::Bool(b) => ValueImpl::Bool(*b),
+                CheckedRef::Address(a) => ValueImpl::Address(*a),
+                CheckedRef::Container(c) => ValueImpl::Container((&*c).clone()),
+            }
+        };
+        Value(v)
     }
 }
 
 impl Reference {
-    pub fn read_ref(self) -> PartialVMResult<Value> {
+    pub fn read_ref(self) -> Value {
```

Comment:
This should be unsafe fn -- you are dereferencing a raw pointer under the hood.

Reply:
Yeah its a tricky one... I suppose I should... but it will make lots of other things unsafe.